<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getGenesFromKEGG</title>
  <meta name="keywords" content="getGenesFromKEGG">
  <meta name="description" content="getGenesFromKEGG">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getGenesFromKEGG.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external/kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getGenesFromKEGG
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getGenesFromKEGG</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function model=getGenesFromKEGG(keggPath,koList) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getGenesFromKEGG
   Retrieves information on all genes stored in KEGG database

   Input:
   keggPath    if keggGenes.mat is not in the RAVEN\external\kegg
               directory, this function will attempt to read data from a
               local FTP dump of the KEGG database. keggPath is the path
               to the root of this database
   koList      the number of genes in KEGG is very large. koList can be a
               cell array with KO identifiers, in which case only genes
               belonging to one of those KEGG orthologies are retrieved
               (opt, default all KOs with associated reactions)

   Output:
   model       a model structure generated from the database. The
               following fields are filled
       id              'KEGG'
       description     'Automatically generated from KEGG database'
       rxns            KO ids
       rxnNames        Name for each entry
       genes           IDs for all the genes. Genes are saved as organism
                       abbreviation:id (same as in KEGG). 'HSA:124' for
                       example is alcohol dehydrogenase in Homo sapiens
       rxnGeneMat      A binary matrix that indicates whether a specific
                       gene is present in a KO id

   NOTE: If the file keggGenes.mat is in the RAVEN\external\kegg directory
   it will be loaded instead of parsing of the KEGG files. If it does not
   exist it will be saved after parsing of the KEGG files. In general, you
   should remove the keggGenes.mat file if you want to rebuild the model
   structure from a newer version of KEGG.

   Usage: model=getGenesFromKEGG(keggPath,koList)

   Simonas Marcisauskas, 2019-09-09


 NOTE: This is how one entry looks in the file

 ENTRY       K11440                      KO
 NAME        gbsB
 DEFINITION  choline dehydrogenase [EC:1.1.1.1]
 PATHWAY     ko00260  Glycine, serine and threonine metabolism
             ko01100  Metabolic pathways
 MODULE      M00555  Betaine biosynthesis, choline =&gt; betaine
 BRITE       KEGG Orthology (KO) [BR:ko00001]
             (list truncated)
 DBLINKS     RN: R08557 R08558
             COG: COG1454
 GENES       BSU: BSU31050(gbsB)
             BSR: I33_3189
             (list truncated)
 REFERENCE   PMID:8752328
   AUTHORS   Boch J, Kempf B, Schmid R, Bremer E
   TITLE     Synthesis of the osmoprotectant glycine... (truncated)
   JOURNAL   J Bacteriol 178:5121-9 (1996)
             DOI:10.1128/JB.178.17.5121-5129.1996
   SEQUENCE  [bsu:BSU31050]
 ///

 The file is not tab-delimited. Instead each label is 12 characters
 (except for '///').

 Check if the genes have been parsed before and saved. If so, load the
 model.</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getModelFromKEGG.html" class="code" title="function [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>	getModelFromKEGG</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function allKOs=getAllKOs(keggPath)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=getGenesFromKEGG(keggPath,koList)</a>
0002 <span class="comment">% getGenesFromKEGG</span>
0003 <span class="comment">%   Retrieves information on all genes stored in KEGG database</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   Input:</span>
0006 <span class="comment">%   keggPath    if keggGenes.mat is not in the RAVEN\external\kegg</span>
0007 <span class="comment">%               directory, this function will attempt to read data from a</span>
0008 <span class="comment">%               local FTP dump of the KEGG database. keggPath is the path</span>
0009 <span class="comment">%               to the root of this database</span>
0010 <span class="comment">%   koList      the number of genes in KEGG is very large. koList can be a</span>
0011 <span class="comment">%               cell array with KO identifiers, in which case only genes</span>
0012 <span class="comment">%               belonging to one of those KEGG orthologies are retrieved</span>
0013 <span class="comment">%               (opt, default all KOs with associated reactions)</span>
0014 <span class="comment">%</span>
0015 <span class="comment">%   Output:</span>
0016 <span class="comment">%   model       a model structure generated from the database. The</span>
0017 <span class="comment">%               following fields are filled</span>
0018 <span class="comment">%       id              'KEGG'</span>
0019 <span class="comment">%       description     'Automatically generated from KEGG database'</span>
0020 <span class="comment">%       rxns            KO ids</span>
0021 <span class="comment">%       rxnNames        Name for each entry</span>
0022 <span class="comment">%       genes           IDs for all the genes. Genes are saved as organism</span>
0023 <span class="comment">%                       abbreviation:id (same as in KEGG). 'HSA:124' for</span>
0024 <span class="comment">%                       example is alcohol dehydrogenase in Homo sapiens</span>
0025 <span class="comment">%       rxnGeneMat      A binary matrix that indicates whether a specific</span>
0026 <span class="comment">%                       gene is present in a KO id</span>
0027 <span class="comment">%</span>
0028 <span class="comment">%   NOTE: If the file keggGenes.mat is in the RAVEN\external\kegg directory</span>
0029 <span class="comment">%   it will be loaded instead of parsing of the KEGG files. If it does not</span>
0030 <span class="comment">%   exist it will be saved after parsing of the KEGG files. In general, you</span>
0031 <span class="comment">%   should remove the keggGenes.mat file if you want to rebuild the model</span>
0032 <span class="comment">%   structure from a newer version of KEGG.</span>
0033 <span class="comment">%</span>
0034 <span class="comment">%   Usage: model=getGenesFromKEGG(keggPath,koList)</span>
0035 <span class="comment">%</span>
0036 <span class="comment">%   Simonas Marcisauskas, 2019-09-09</span>
0037 <span class="comment">%</span>
0038 <span class="comment">%</span>
0039 <span class="comment">% NOTE: This is how one entry looks in the file</span>
0040 <span class="comment">%</span>
0041 <span class="comment">% ENTRY       K11440                      KO</span>
0042 <span class="comment">% NAME        gbsB</span>
0043 <span class="comment">% DEFINITION  choline dehydrogenase [EC:1.1.1.1]</span>
0044 <span class="comment">% PATHWAY     ko00260  Glycine, serine and threonine metabolism</span>
0045 <span class="comment">%             ko01100  Metabolic pathways</span>
0046 <span class="comment">% MODULE      M00555  Betaine biosynthesis, choline =&gt; betaine</span>
0047 <span class="comment">% BRITE       KEGG Orthology (KO) [BR:ko00001]</span>
0048 <span class="comment">%             (list truncated)</span>
0049 <span class="comment">% DBLINKS     RN: R08557 R08558</span>
0050 <span class="comment">%             COG: COG1454</span>
0051 <span class="comment">% GENES       BSU: BSU31050(gbsB)</span>
0052 <span class="comment">%             BSR: I33_3189</span>
0053 <span class="comment">%             (list truncated)</span>
0054 <span class="comment">% REFERENCE   PMID:8752328</span>
0055 <span class="comment">%   AUTHORS   Boch J, Kempf B, Schmid R, Bremer E</span>
0056 <span class="comment">%   TITLE     Synthesis of the osmoprotectant glycine... (truncated)</span>
0057 <span class="comment">%   JOURNAL   J Bacteriol 178:5121-9 (1996)</span>
0058 <span class="comment">%             DOI:10.1128/JB.178.17.5121-5129.1996</span>
0059 <span class="comment">%   SEQUENCE  [bsu:BSU31050]</span>
0060 <span class="comment">% ///</span>
0061 <span class="comment">%</span>
0062 <span class="comment">% The file is not tab-delimited. Instead each label is 12 characters</span>
0063 <span class="comment">% (except for '///').</span>
0064 <span class="comment">%</span>
0065 <span class="comment">% Check if the genes have been parsed before and saved. If so, load the</span>
0066 <span class="comment">% model.</span>
0067 <span class="comment">%</span>
0068 
0069 <span class="keyword">if</span> nargin&lt;1
0070     keggPath=<span class="string">'RAVEN/external/kegg'</span>;
0071 <span class="keyword">end</span>
0072 
0073 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0074 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0075 genesFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggGenes.mat'</span>);
0076 <span class="keyword">if</span> exist(genesFile, <span class="string">'file'</span>)
0077     fprintf([<span class="string">'Importing KEGG genes from '</span> strrep(genesFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'... '</span>]);
0078     load(genesFile);
0079 <span class="keyword">else</span>
0080     fprintf([<span class="string">'NOTE: Cannot locate '</span> strrep(genesFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">', it will therefore be generated from the local KEGG database\n'</span>]);
0081     <span class="keyword">if</span> ~exist(fullfile(keggPath,<span class="string">'ko'</span>),<span class="string">'file'</span>) || ~exist(fullfile(keggPath,<span class="string">'reaction'</span>),<span class="string">'file'</span>)
0082         EM=fprintf([<span class="string">'The files ''ko'' and ''reaction'' cannot be located at '</span> strrep(keggPath,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'/ and should be downloaded from the KEGG FTP\n'</span>]);
0083         dispEM(EM);
0084     <span class="keyword">else</span>
0085         fprintf(<span class="string">'Generating keggGenes.mat file... '</span>);
0086         <span class="comment">%Get all KOs that are associated to reactions</span>
0087         allKOs=<a href="#_sub1" class="code" title="subfunction allKOs=getAllKOs(keggPath)">getAllKOs</a>(keggPath);
0088         
0089         <span class="comment">%Since the list of genes will be accessed many times the hash table</span>
0090         <span class="comment">%is established</span>
0091         geneMap=containers.Map(<span class="string">'KeyType'</span>,<span class="string">'char'</span>,<span class="string">'ValueType'</span>,<span class="string">'int32'</span>);
0092         
0093         <span class="comment">%Add new functionality in the order specified in models</span>
0094         model.id=<span class="string">'KEGG'</span>;
0095         model.description=<span class="string">'Automatically generated from KEGG database'</span>;
0096         
0097         <span class="comment">%Preallocate memory</span>
0098         model.rxns=cell(numel(allKOs),1);
0099         model.rxnNames=cell(numel(allKOs),1);
0100         
0101         <span class="comment">%Reserve space for 500000 genes</span>
0102         model.genes=cell(500000,1);
0103         
0104         <span class="comment">%Load information on KO ID, name, and associated genes</span>
0105         fid = fopen(fullfile(keggPath,<span class="string">'ko'</span>), <span class="string">'r'</span>);
0106         
0107         <span class="comment">%Keeps track of how many KOs that have been added</span>
0108         koCounter=0;
0109         nGenes=0;
0110         addingGenes=false;
0111         
0112         <span class="comment">%These contain the mapping between KOs and genes and are used to</span>
0113         <span class="comment">%generate the model.rxnGeneMat in the end</span>
0114         koIndex=zeros(5000000,1);
0115         geneIndex=koIndex;
0116         nMappings=0;
0117         
0118         skipRead=false;
0119         
0120         <span class="comment">%Loop through the file</span>
0121         <span class="keyword">while</span> 1
0122             <span class="comment">%Get the next line</span>
0123             <span class="keyword">if</span> skipRead==false
0124                 tline = fgetl(fid);
0125             <span class="keyword">else</span>
0126                 skipRead=false;
0127             <span class="keyword">end</span>
0128             
0129             <span class="comment">%Abort at end of file</span>
0130             <span class="keyword">if</span> ~ischar(tline)
0131                 <span class="keyword">break</span>;
0132             <span class="keyword">end</span>
0133             
0134             <span class="comment">%Skip '///'</span>
0135             <span class="keyword">if</span> numel(tline)&lt;12
0136                 addingGenes=false;
0137                 <span class="keyword">continue</span>;
0138             <span class="keyword">end</span>
0139             
0140             <span class="comment">%Check if it's a new reaction</span>
0141             <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'ENTRY       '</span>)
0142                 <span class="comment">%Check if it should be added</span>
0143                 koID=tline(13:18);
0144                 
0145                 <span class="keyword">if</span> ismember(koID,allKOs)
0146                     addMe=true;
0147                     koCounter=koCounter+1;
0148                 <span class="keyword">else</span>
0149                     addMe=false;
0150                     <span class="keyword">continue</span>;
0151                 <span class="keyword">end</span>
0152                 
0153                 <span class="comment">%Add reaction ID (always 6 characters)</span>
0154                 model.rxns{koCounter}=koID;
0155                 
0156                 model.rxnNames{koCounter}=<span class="string">''</span>;
0157                 <span class="comment">%Will be overwritten if it exists</span>
0158             <span class="keyword">end</span>
0159             
0160             <span class="comment">%To get here we must be in a KO that should be added</span>
0161             <span class="keyword">if</span> addMe==true
0162                 <span class="comment">%Add name</span>
0163                 <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'DEFINITION  '</span>)
0164                     model.rxnNames{koCounter}=tline(13:end);
0165                 <span class="keyword">end</span>
0166                 
0167                 <span class="comment">%Check if it's time to start adding genes</span>
0168                 <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'GENES       '</span>)
0169                     addingGenes=true;
0170                 <span class="keyword">end</span>
0171                 
0172                 <span class="comment">%Add genes</span>
0173                 <span class="keyword">if</span> addingGenes==true
0174                     <span class="comment">%We are now adding genes for the current KO. All gene</span>
0175                     <span class="comment">%rows start with 12 spaces. If this is not true it</span>
0176                     <span class="comment">%means that there is additional annotation such as</span>
0177                     <span class="comment">%AUTHORS. This is not common but it exists</span>
0178                     <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'            '</span>) || strcmp(tline(1:12),<span class="string">'GENES       '</span>)
0179                         geneLine=tline(13:end);
0180                         
0181                         <span class="comment">%Check if the line is from a new organism of from</span>
0182                         <span class="comment">%the same as before</span>
0183                         <span class="keyword">if</span> strcmp(geneLine(4),<span class="string">':'</span>)
0184                             <span class="comment">%If organism id contains 3 letters;</span>
0185                             currentOrganism=geneLine(1:3);
0186                             <span class="comment">%Parse the string</span>
0187                             genes=regexp(geneLine(6:end),<span class="string">' '</span>,<span class="string">'split'</span>);
0188                             genes=strcat(currentOrganism,<span class="string">':'</span>,genes(:));
0189                         <span class="keyword">elseif</span> strcmp(geneLine(5),<span class="string">':'</span>)
0190                             <span class="comment">%If organism id contains 4 letters;</span>
0191                             currentOrganism=geneLine(1:4);
0192                             <span class="comment">%Parse the string</span>
0193                             genes=regexp(geneLine(7:end),<span class="string">' '</span>,<span class="string">'split'</span>);
0194                             genes=strcat(currentOrganism,<span class="string">':'</span>,genes(:));
0195                         <span class="keyword">end</span>
0196                         
0197                         <span class="comment">%Add the genes to the gene list</span>
0198                         <span class="keyword">for</span> i=1:numel(genes)
0199                             geneString=genes{i};
0200                             <span class="keyword">if</span> geneMap.isKey(geneString)
0201                                 index=geneMap(geneString);
0202                             <span class="keyword">else</span>
0203                                 nGenes=nGenes+1;
0204                                 model.genes{nGenes}=geneString;
0205                                 index=nGenes;
0206                                 geneMap(geneString)=index;
0207                             <span class="keyword">end</span>
0208                             
0209                             <span class="comment">%Add the connection to the KO</span>
0210                             nMappings=nMappings+1;
0211                             koIndex(nMappings)=koCounter;
0212                             geneIndex(nMappings)=index;
0213                         <span class="keyword">end</span>
0214                     <span class="keyword">else</span>
0215                         <span class="comment">%Now we want to throw away everything until the</span>
0216                         <span class="comment">%next entry</span>
0217                         <span class="keyword">while</span> 1
0218                             tline = fgetl(fid);
0219                             <span class="keyword">if</span> strcmp(tline,<span class="string">'///'</span>)
0220                                 <span class="comment">%When the new entry is found, skip reading</span>
0221                                 <span class="comment">%one line to fit with the rest of the code</span>
0222                                 skipRead=true;
0223                                 <span class="keyword">break</span>;
0224                             <span class="keyword">end</span>
0225                         <span class="keyword">end</span>
0226                     <span class="keyword">end</span>
0227                 <span class="keyword">end</span>
0228             <span class="keyword">end</span>
0229         <span class="keyword">end</span>
0230         <span class="comment">%Close the file</span>
0231         fclose(fid);
0232         
0233         <span class="comment">%If too much space was allocated, shrink the model</span>
0234         model.rxns=model.rxns(1:koCounter);
0235         model.rxnNames=model.rxnNames(1:koCounter);
0236         model.genes=model.genes(1:nGenes);
0237         
0238         <span class="comment">%Retrieve and add the gene associations</span>
0239         model.rxnGeneMat=sparse(koIndex(1:nMappings),geneIndex(1:nMappings),ones(nMappings,1));
0240         
0241         <span class="comment">%To make sure the size is correct if the last KOs don't have genes</span>
0242         <span class="keyword">if</span> size(model.rxnGeneMat,1)~=koCounter
0243             model.rxnGeneMat(koCounter,1)=0;
0244         <span class="keyword">end</span>
0245         
0246         <span class="comment">%Trim the genes so that they only contain information that can be</span>
0247         <span class="comment">%matched to the KEGG file of protein sequences (remove all</span>
0248         <span class="comment">%information after first parenthesis)</span>
0249         <span class="comment">%NOTE: For some reason the organism abbreviation should be in lower</span>
0250         <span class="comment">%case in this database. Fix this here</span>
0251         <span class="keyword">for</span> i=1:numel(model.genes)
0252             parIndex=strfind(model.genes{i},<span class="string">'('</span>);
0253             <span class="keyword">if</span> any(parIndex)
0254                 model.genes{i}=model.genes{i}(1:parIndex-1);
0255             <span class="keyword">end</span>
0256             colIndex=strfind(model.genes{i},<span class="string">':'</span>);
0257             model.genes{i}=[lower(model.genes{i}(1:colIndex-1)) model.genes{i}(colIndex:end)];
0258         <span class="keyword">end</span>
0259         
0260         <span class="comment">%Save the model structure</span>
0261         save(genesFile,<span class="string">'model'</span>);
0262     <span class="keyword">end</span>
0263 <span class="keyword">end</span>
0264 
0265 <span class="comment">%Only get the KOs in koList</span>
0266 I=~ismember(model.rxns,koList);
0267 model=removeReactions(model,I,true,true);
0268 fprintf(<span class="string">'COMPLETE\n'</span>);
0269 <span class="keyword">end</span>
0270 
0271 <a name="_sub1" href="#_subfunctions" class="code">function allKOs=getAllKOs(keggPath)</a>
0272 <span class="comment">%Retrieves all KOs that are associated to reactions. This is because the</span>
0273 <span class="comment">%number of genes in KEGG is very large so without this parsing it would</span>
0274 <span class="comment">%take many hours</span>
0275 
0276 allKOs={};
0277 
0278 <span class="comment">%First check if the reactions have already been parsed</span>
0279 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0280 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0281 rxnsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggRxns.mat'</span>);
0282 <span class="keyword">if</span> exist(rxnsFile, <span class="string">'file'</span>)
0283     fprintf([<span class="string">'NOTE: Importing KEGG ORTHOLOGY list from '</span> strrep(rxnsFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0284     load(rxnsFile,<span class="string">'model'</span>);
0285     
0286     <span class="comment">%Loop through the reactions and add the corresponding genes</span>
0287     <span class="keyword">for</span> i=1:numel(model.rxns)
0288         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0289             <span class="comment">%Get all KOs</span>
0290             allKOs=[allKOs;model.rxnMiriams{i}.value(strcmpi(model.rxnMiriams{i}.name,<span class="string">'kegg.orthology'</span>))];
0291         <span class="keyword">end</span>
0292     <span class="keyword">end</span>
0293 <span class="keyword">else</span>
0294     <span class="comment">%Parse the reaction file instead First load information on reaction ID,</span>
0295     <span class="comment">%reaction name, KO, pathway and ec-number</span>
0296     fid = fopen(fullfile(keggPath,<span class="string">'reaction'</span>), <span class="string">'r'</span>);
0297     orthology=false;
0298     <span class="keyword">while</span> 1
0299         <span class="comment">%Get the next line</span>
0300         tline = fgetl(fid);
0301         
0302         <span class="comment">%Abort at end of file</span>
0303         <span class="keyword">if</span> ~ischar(tline)
0304             <span class="keyword">break</span>;
0305         <span class="keyword">end</span>
0306         
0307         <span class="comment">%Skip '///'</span>
0308         <span class="keyword">if</span> numel(tline)&lt;12
0309             <span class="keyword">continue</span>;
0310         <span class="keyword">end</span>
0311         
0312         <span class="comment">%Check if it's a new reaction</span>
0313         <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'ENTRY       '</span>)
0314             orthology=false;
0315         <span class="keyword">end</span>
0316         
0317         <span class="keyword">if</span> strcmp(tline(1:9),<span class="string">'REFERENCE'</span>)
0318             orthology=false;
0319         <span class="keyword">end</span>
0320         
0321         <span class="comment">%Add KO-ids</span>
0322         <span class="keyword">if</span> numel(tline)&gt;16
0323             <span class="keyword">if</span> strcmp(tline(1:16),<span class="string">'ORTHOLOGY   KO: '</span>) || strcmp(tline(1:16),<span class="string">'            KO: '</span>) || strcmp(tline(1:12),<span class="string">'ORTHOLOGY   '</span>) || orthology==true
0324                 <span class="keyword">if</span> strcmp(tline(13:16),<span class="string">'KO:'</span>)
0325                     <span class="comment">%This is in the old version</span>
0326                     allKOs=[allKOs;tline(17:22)];
0327                 <span class="keyword">elseif</span> strcmp(tline(1:12),<span class="string">'ORTHOLOGY   '</span>)
0328                     <span class="comment">%This means that it found one KO in the new format and</span>
0329                     <span class="comment">%that subsequent lines might be other KOs</span>
0330                     orthology=true;
0331                     allKOs=[allKOs;tline(13:18)];
0332                 <span class="keyword">end</span>
0333             <span class="keyword">end</span>
0334         <span class="keyword">end</span>
0335     <span class="keyword">end</span>
0336     
0337     <span class="comment">%Close the file</span>
0338     fclose(fid);
0339 <span class="keyword">end</span>
0340 allKOs=unique(allKOs);
0341 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 03-Jun-2020 13:07:00 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>